src/Modules/UngradedModules.jl

```diff
@@ -807,16 +908,211 @@ function base_ring(M::SubModuleOfFreeModule)
+end
+
+@doc Markdown.doc"""
+    reduced_groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+
+Compute a reduced Gröbner basis with respect to the given `ordering`. The return type is `ModuleGens`.
+"""
+function reduced_groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  @assert is_global(ordering)
+  
+  if !haskey(submod.groebner_basis, ordering)
+    gb = compute_std_basis(submod, ordering, true)
+    submod.groebner_basis[ordering] = gb
+    return gb
+  else
+    if !(submod.groebner_basis[ordering].is_reduced)
+      red_gb = get_attribute(submod.groebner_basis[ordering], :reduced_groebner_basis)
+      if red_gb === nothing
+        red_gb = compute_std_basis(submod, ordering, true)
+        set_attribute!(submod.groebner_basis[ordering], :reduced_groebner_basis => red_gb)
+      else
+        return red_gb
+      end
+    else
+      return submod.groebner_basis[ordering]
+    end
+  end
+end
+
+function leading_module(submod::SubModuleOfFreeModule)
+  return leading_module(submod, default_ordering(submod))
+end
+
+function leading_module(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  gb = std_basis(submod, ordering)
+  return SubModuleOfFreeModule(submod.F, leading_monomials(gb))
+end
+
+@doc Markdown.doc"""
+    compute_std_basis(submod::SubModuleOfFreeModule)
+
+Compute a standard basis of `submod` and return it as a `ModuleGens`.
+"""
+function compute_std_basis(submod::SubModuleOfFreeModule)
+  return compute_std_basis(submod, default_ordering(submod))
+end
+
+@doc Markdown.doc"""
+    compute_std_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering, reduced::Bool=false)
+
+Compute a stanard basis of `submod` with respect to the given ordering.
+Allowed orderings are those that are allowed as orderings for Singular polynomial rings.
+In case `reduced` is `true` and the ordering is global, a reduced Gröbner basis is computed.
+"""
+function compute_std_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering, reduced::Bool=false)
+  if reduced
+    @assert is_global(ordering)
+  end
+  mg = ModuleGens(oscar_generators(submod.gens), submod.F , ordering)
+  gb = std_basis(mg, reduced)
+  oscar_assure(gb)
+  gb.isGB = true
+  gb.S.isGB = true
+  gb.ordering = ordering
+  gb.is_reduced = reduced
+  return gb
+end
+
+function show_gb(io::IO, GB::ModuleGens)
+  if isdefined(GB, :quo_GB)
+    show_relative_groebner_basis(io, GB, GB.quo_GB, GB.is_reduced)
+  else
+    show_non_relative_groebner_basis(io, GB, GB.is_reduced)
+  end
+end
+
+function show_groebner_basis_helper(io::IO, sub::ModuleGens, init::String)
+  F = sub.F
+
+  print(io, init)
+
+  # This shall be removed when Oscar orderings are used
+  singular_assure(sub)
+  vectors = [[] for _ in 1:length(sub.O)]
+  for i in 1:length(sub.O)
+    monomials = []
+    v = sub.S[i]
+    l = Singular.lead(v)
+    while !iszero(l)
+      push!(monomials, l)
+      v = v - l
+      l = Singular.lead(v)
+    end
```

Comment:
Is there really no better method to iterate over the monomials (? terms?) here? 

A brief check reveals that one can iterate over `svector`:
```
julia> f = M[1]
x*y*gen(2)+x*gen(1)+y*gen(3)+gen(2)+gen(1)

julia> collect(f)
5-element Vector{Any}:
 (2, [1, 1], 1)
 (1, [1, 0], 1)
 (3, [0, 1], 1)
 (2, [0, 0], 1)
 (1, [0, 0], 1)
```
But maybe improving this is not worth the effort if you plan to do something else soon anyway?

Reply:
Maybe it's worth the effort, since probably it's better (performance-wise) to stay on the Singular side when printing. But I don't know if it's better to use these triples instead of calling `lead`.

Reply:
The problem is not `lead` which is fast, it is the repeated subtraction of the lead term, creating ~2N new polynomials to iterate over N terms. While the internal datastructure really allows for fast iteration over all terms

Reply:
Was just meant as a workaround, but we should still improve it for use in the case there is only a Singular side in the ModuleGens.

Reply:
If it's just a temporary workaround, fine by me. It'd be great if  you could annotate all such temporary workarounds with a very explicit message, ideally involving a keyword like "TODO" or "FIXME" so it can quickly be found again in the future.

Reply:
Okay, I understand. Probably it would be the best solution if there was a `monomials`-function in Singular.jl...